# Hypothesis testing problems

In [1]:
import numpy as np
from scipy.stats import levene, shapiro
from scipy.stats import f_oneway
from statsmodels.stats.multicomp import pairwise_tukeyhsd


## Exercise 1

You are a nutritionist researching two different types of diets to see if there is a significant difference in weight loss after one month. You choose two random groups of people; one group follows the first diet, and the other follows the second. At the end of the month, you record the weight loss (in kg) for each person:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

With these data, it seeks to answer the following question: Is there a significant difference in average weight loss between people who followed the first diet and those who followed the second diet?

To draw conclusions, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the test to test the hypothesis. You can use a Student's t-test.
- Analyze the conclusions.

## Formulating both hypothesis

In [2]:
H0 = "There is no difference in the population means (average weight loss) between Diet A and Diet B."
H1 = "There is a difference in the population means (average weight loss) between Diet A and Diet B."

## Testing the hypothesis
> 1. Diet 1 and Diet 2 weight loss data
> 2. Performing two-sample t-test
> 3. Interpretation

In [3]:
# 1
diet1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
diet2 = [3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2]

from scipy.stats import ttest_ind

# 2
t_stat, p_value = ttest_ind(diet1, diet2, equal_var=True)

print(f"t-statistic: {t_stat:.3f}")
print(f"p-value: {p_value:.3f}")

# 3
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in average weight loss between Diet 1 and Diet 2.")
else:
    print("Fail to reject the null hypothesis: No significant difference in average weight loss between Diet 1 and Diet 2.")
    

t-statistic: -3.538
p-value: 0.004
Reject the null hypothesis: There is a significant difference in average weight loss between Diet 1 and Diet 2.


## ANOVA

**ANOVA** (*Analysis of Variance*) is a statistical technique used to compare the measures of two or more groups. The idea behind ANOVA is to decompose the total variability in the data into two components: between-group variability and within-group variability:

- **Between-group variability**: This variability refers to the differences between the group means. If this variability is considerably larger than the within-group variability, it could be an indication that at least one of the group means is different.
- **Within-group variability**: This variability refers to the dispersion of the data within each group. If all groups have similar variability, then any noticeable difference in group means could be considered significant.

Hypotheses in ANOVA typically include:

- **Null hypothesis** ($H₀$): The means of all groups are equal.
- **Alternative hypothesis** ($H₁$): At least one of the group means is different.

If the ANOVA test result is significant (e.g., a p-value less than a threshold such as 0.05), this suggests that at least one group mean is different.

## Exercise 2

A farmer decides to test three different types of fertilizers to determine if one is superior in terms of corn production. The farmer plants corn on 15 identical plots and uses all three fertilizers (5 plots for each type). At the end of the season, he measures the corn yield (in kg) of each plot, with the following result:

| Fertilizer 1 | Fertilizer 2 | Fertilizer 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

With this data, he seeks to answer the following question: Is there a significant difference in average corn yield between the three types of fertilizers?

To help you, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the ANOVA test.
- Analyze the conclusions.
- If one fertilizer is better than another, how can we know it?

### 1. State the hypotheses

- **Null hypothesis** ($H_0$): The mean corn yields for all three fertilizers are equal.
- **Alternative hypothesis** ($H_1$): At least one fertilizer has a different mean corn yield.

### 2. Prepare the data

Let's organize the corn yield data for each fertilizer.

In [4]:
# Corn yield data for each fertilizer
fertilizer_1 = [20, 21, 20, 19, 20]
fertilizer_2 = [22, 21, 23, 22, 21]
fertilizer_3 = [24, 23, 22, 23, 24]

### 3. Check ANOVA assumptions

Before performing ANOVA, we need to check assumptions: normality and equal variances.

In [5]:
# Test for equal variances (Levene's test)
levene_stat, levene_p = levene(fertilizer_1, fertilizer_2, fertilizer_3)
print(f"Levene's test - Statistic: {levene_stat:.3f}, p-value: {levene_p:.4f}")

# Test for normality (Shapiro-Wilk test)
for i, fert in enumerate([fertilizer_1, fertilizer_2, fertilizer_3], 1):
    stat, p = shapiro(fert)
    print(f"Shapiro test Fertilizer {i} - Statistic: {stat:.3f}, p-value: {p:.4f}")

# Interpretation
if levene_p > 0.05:
    print("\nEqual variances assumption met (Levene's test p > 0.05)")
else:
    print("\nEqual variances assumption violated (Levene's test p < 0.05)")

Levene's test - Statistic: 0.222, p-value: 0.8040
Shapiro test Fertilizer 1 - Statistic: 0.883, p-value: 0.3254
Shapiro test Fertilizer 2 - Statistic: 0.881, p-value: 0.3140
Shapiro test Fertilizer 3 - Statistic: 0.881, p-value: 0.3140

Equal variances assumption met (Levene's test p > 0.05)


### 4. Perform the ANOVA test

Since assumptions are checked, we can proceed with one-way ANOVA.

In [6]:
# Perform one-way ANOVA
f_stat, p_value = f_oneway(fertilizer_1, fertilizer_2, fertilizer_3)

print(f"F-statistic: {f_stat:.3f}")
print(f"p-value: {p_value:.4f}")

# Set significance level
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in mean corn yield between at least two fertilizers.")
else:
    print("Fail to reject the null hypothesis: No significant difference in mean corn yield between the fertilizers.")

F-statistic: 20.316
p-value: 0.0001
Reject the null hypothesis: There is a significant difference in mean corn yield between at least two fertilizers.


### 5. Conclusion

If the p-value is less than 0.05, we reject the null hypothesis and conclude that at least one fertilizer produces a different average yield. If not, we conclude there is no significant difference.

### 6. Identifying the best fertilizer (Post-hoc analysis)

With the fact that ANOVA test was significant, we can use a post-hoc test (such as Tukey's HSD) to determine which fertilizers differ from each other.

In [7]:
# Combine data and create group labels
data = np.array(fertilizer_1 + fertilizer_2 + fertilizer_3)
groups = (["Fertilizer 1"] * len(fertilizer_1) +
		  ["Fertilizer 2"] * len(fertilizer_2) +
		  ["Fertilizer 3"] * len(fertilizer_3))

# Tukey's HSD test
tukey = pairwise_tukeyhsd(endog=data, groups=groups, alpha=0.05)
print(tukey)

     Multiple Comparison of Means - Tukey HSD, FWER=0.05      
   group1       group2    meandiff p-adj  lower  upper  reject
--------------------------------------------------------------
Fertilizer 1 Fertilizer 2      1.8 0.0099 0.4572 3.1428   True
Fertilizer 1 Fertilizer 3      3.2 0.0001 1.8572 4.5428   True
Fertilizer 2 Fertilizer 3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------------------


### 7. Interpretation

The Tukey HSD results are as follows:

| Comparison                  | Mean Diff | p-adj  | 95% CI Lower | 95% CI Upper | Significant |
|-----------------------------|-----------|--------|--------------|--------------|-------------|
| Fertilizer 1 vs Fertilizer 2|    1.8    | 0.0099 |    0.4572    |    3.1428    |    Yes      |
| Fertilizer 1 vs Fertilizer 3|    3.2    | 0.0001 |    1.8572    |    4.5428    |    Yes      |
| Fertilizer 2 vs Fertilizer 3|    1.4    | 0.0409 |    0.0572    |    2.7428    |    Yes      |

All pairwise comparisons are significant at the 0.05 level.

## Conclusion

Based on the Tukey HSD results, **Fertilizer 3** is the best fertilizer. It has the highest mean yield and is significantly different from both Fertilizer 1 and Fertilizer 2 (all pairwise comparisons are significant at the 0.05 level).